craft_platforms/_charmcraft.py

```diff
@@ -0,0 +1,77 @@
+# This file is part of craft-platforms.
+#
+# Copyright 2024 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify it
+# under the terms of the GNU Lesser General Public License version 3, as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranties of MERCHANTABILITY,
+# SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License along
+# with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""Charmcraft-specific platforms information."""
+
+import itertools
+from collections.abc import Sequence
+
+from craft_platforms import _architectures, _buildinfo, _distro, _platforms
+
+
+def get_platforms_charm_build_plan(
+    base: str,
+    platforms: _platforms.Platforms | None,
+    build_base: str | None = None,
+) -> Sequence[_buildinfo.BuildInfo]:
+    """Generate the build plan for a platforms-based charm."""
+    distro_base = _distro.DistroBase.from_str(build_base or base)
+    if platforms is None:
+        # If no platforms are specified, build for all default architectures without
+        # an option of cross-compiling.
+        return [
+            _buildinfo.BuildInfo(
+                platform=arch.value,
+                build_on=arch,
+                build_for=arch,
+                base=distro_base,
+            )
+            for arch in _architectures.CRAFT_DEFAULT_ARCHITECTURES
+        ]
+    build_plan: list[_buildinfo.BuildInfo] = []
+    for platform_name, platform in platforms.items():
+        if platform is None:
+            # This is a workaround for Python 3.10.
+            # In python 3.12+ we can just check:
+            # `if platform_name not in _architectures.DebianArchitecture`
+            try:
+                architecture = _architectures.DebianArchitecture(platform_name)
+            except ValueError:
+                raise ValueError(
+                    f"Platform name {platform_name!r} is not a valid Debian architecture. "
+                    "Specify a build-on and build-for.",
+                ) from None
+            build_plan.append(
+                _buildinfo.BuildInfo(
+                    platform=platform_name,
+                    build_on=architecture,
+                    build_for=architecture,
+                    base=distro_base,
+                ),
+            )
+        else:
+            for build_on, build_for in itertools.product(
```

Comment:
so for charmcraft `build-on: [x, y]` and `build-for: [z, w, k]` will produce 6 build infos?

Reply:
ah I see that there's a limit of 1 in build-for

Reply:
Yeah there's that limitation, but it's implemented this way so we can hopefully basically copy this everywhere. (Note that this version is not the same as how charmcraft does it since the default implementation builds for all default architectures - in practice craft-application will simply filter this and Charmcraft will replace its build planning with this + logic that if the current architecture is not in the default architectures we replace an empty build plan with `{on,for} current_arch`